hibernate-reactive-core/src/main/java/org/hibernate/reactive/mutiny/impl/MutinySessionFactoryImpl.java

```diff
@@ -146,30 +155,46 @@ private ReactiveConnection proxyConnection(String tenantId) {
 		if ( current!=null && current.isOpen() ) {
 			return work.apply( current );
 		}
-		return newSession( tenantId )
-				.chain( session -> {
-					context.put(Mutiny.Session.class, id, session);
-					return work.apply(session)
-							.eventually( () -> {
-								context.remove(Mutiny.Session.class, id);
-								return session.close();
-							} );
-				} );
+		return withSession( Mutiny.Session.class, newSession( tenantId ), Mutiny.Session::close, work, id );
 	}
 
 	@Override
 	public <T> Uni<T> withStatelessSession(Function<Mutiny.StatelessSession, Uni<T>> work) {
-		Mutiny.StatelessSession current = context.get(Mutiny.StatelessSession.class, statelessSessionId);
-		if ( current!=null && current.isOpen() ) {
+		Mutiny.StatelessSession current = context.get( Mutiny.StatelessSession.class, statelessSessionId );
+		if ( current != null && current.isOpen() ) {
 			return work.apply( current );
 		}
-		return newStatelessSession()
+		return withSession( Mutiny.StatelessSession.class, newStatelessSession(), Mutiny.StatelessSession::close, work, statelessSessionId );
+	}
+
+	private<S, T> Uni<T> withSession(
+			Class<S> sessionType,
+			Uni<S> sessionUni,
+			Function<S, Uni<Void>> closeSession,
+			Function<S, Uni<T>> work,
+			String id) {
+		return sessionUni
 				.chain( session -> {
-					context.put(Mutiny.StatelessSession.class, statelessSessionId, session);
-					return work.apply(session).eventually( () -> {
-						context.remove(Mutiny.StatelessSession.class, statelessSessionId);
-						return session.close();
-					} );
+					try {
+						context.put( sessionType, id, session );
+						return work.apply( session )
+								.invoke( result -> context.remove( sessionType, id ) )
+								.eventually( () -> closeSession.apply( session ) );
+					}
+					catch (Exception e) {
+						// Ignoring additional exceptions after this point
+						try {
+							context.remove( sessionType, id );
+							return closeSession.apply( session )
+									.onItemOrFailure()
+									.transformToUni( (unused, throwable) ->  Uni.createFrom().failure( e ) );
+						}
+						catch (Exception removeEx ) {
+							return closeSession.apply( session )
+									.onItemOrFailure()
+									.transformToUni( (unused, throwable) ->  Uni.createFrom().failure( e ) );
+						}
+					}
```

Comment:
Oh so the issue was that even though `work` looks like a `Uni`, it's actually a function that can throw its own exception, that is *not* captured in the stream. That much makes perfect sense. But there seems to be other stuff going on here.

Reply:
I was thinking that `context.remove` might throw an exception. Mostly because it is a service and might reimplemented.

Reply:
I think we could reasonably require that `remove()` not throw exceptions. It's just a cleanup method after all.

Reply:
My philosophy on this would be: a problem that occurs cleaning up resources _after_ a txn has already completed successfully _might_ be something that is of interest to a system administrator (so should be logged) but is almost certainly _not_ something that should be reported to the user. What the user needs to see is that their txn was successful.